File: em_ppca.m
%% Principal Component Analysis Based On Expectation Maximization Algorithm
function [W,sigma_square,Xn,t_mean] = em_ppca(t,k)

% Check dimenality Constraint
if k<1 |  k>size(t,2)
   error('Number of PCs must be integer, >0, <dim');
end

%No of Iteration
iter = 20;

% find height and width of data matrix
[height,width] = size(t); 

% Find mean value of [t1 t2 ... tm]
% Mind mean vector of observed data vectors
t_mean = (sum(t(:,1:width)')')/width; 

% Normalize Data matrix
t = t - (t_mean)*ones(1,width);

% Initialy W and sigma^2 will be randomly selected.
W = randn(height,k);
sigma_square = randn(1,1);

disp('EM Algorithm running...');
for i=1:iter
    % Find M = W'W + Sigma^2*I
    M = W'*W + sigma_square*eye(k,k); 
    
    % Find inverse of M
    In_M = inv(M);     
    
    % Expected Xn
    x = (In_M)*W'*t; 
    
    % Find Expected of XnXn'
    xn_xn_T = sigma_square*(In_M) + x*x'; 
    
    % Take Old value of W
    old_W = W;
    
    % Take new value of W
    W = t*x'*inv(xn_xn_T); 
    
    sigma_square = 0;
    for i=1:width
        sigma_square=sigma_square + (norm(t(:,1)-t_mean)^2 - 2*((In_M)*(old_W')*(t(:,i)-t_mean))'*(W')*(t(:,i)-t_mean));
    end
    sigma_square = sigma_square + trace(xn_xn_T*(W')*W);
    
    sigma_square = sigma_square/(width*height);
end

disp('EM Algorithm completed. W is created. Press enter to continue...');
pause;

% Find M = W'W + sigma^2*I
M = W'*W + sigma_square * eye(k,k);
% Inverse of M
In_M = inv(M); 
% Latent Variable Xn
Xn = W'*t; 
disp('Principal Component are ready...press enter to continue...');
pause;

end

File: GM_with_ppca/em_ppca.py
import numpy as np
import math as mt
from numpy.linalg import inv
from numpy import linalg as LA

def em_ppca(data,no_pc,max_iter,error):

	# Data  : for which we want to find principal components
	# no_pc : how many principal components use wants
	#		  -> condition no_pc < no of data vectors (columns)
	# max_iter : maximum iteration that user allow to run EM algorithm
	# error : Mean Squared Error that user allows	

	# Take height and width of data matrix
	data_size = data.shape
	height = data_size[0]
	width = data_size[1]

	# Check dimensionality constraint
	if no_pc<1 or no_pc>width:
		print("Oops!  That was no valid number.  Try again...")
		return(0)

	# Find mean value of given matrix
	t_mean = np.zeros((height, 1),dtype='float64')
	for i in range(0,height):
		for j in range(0,width):
			if data[i][j] is not None:
				t_mean[i][0] = t_mean[i][0] + data[i][j]
	t_mean = t_mean/height

	# Normalize Data Matrix
	for i in range(0,height):
		for j in range(0,width):
			if data[i][j] is not None:
				data[i][j] = data[i][j] - t_mean[i][0];
			else:
				data[i][j] = 0;

	# Initialy W and sigma^2 will be randomly selected.
	mu,sigma = 0,1
	print('W and Sigma square are initialized')
	W = np.random.normal(mu, sigma, (height,no_pc))
	sigma_square = np.random.normal(mu,sigma)

	print('EM Algorithm is running')
	for i in range(0,max_iter):
		#Find W transpose
		W_T = W.transpose()
		#Find M = W'W + Sigma^2*I
		M = np.dot(W_T,W) + (sigma_square*np.identity(no_pc))
		# Find Inverse of Matrix
		In_M = inv(M)
		# Expected Xn
		Xn = np.zeros((no_pc, width),dtype='float64')
		Xn_Xn_T = np.zeros((no_pc, no_pc),dtype='float64')

		for j in range(0,width):
			Xn[:,j] = np.dot(np.dot(In_M,W_T),data[:,j])
			Xn_Xn_T = Xn_Xn_T + ((sigma_square*In_M)+(np.dot(Xn[:,j].reshape(len(Xn),1),(Xn[:,j].reshape(len(Xn),1).transpose()))))

		old_W = W

		temp1 = np.zeros((height, no_pc),dtype='float64')
		for j in range(0,width):
			temp1 = temp1 + (np.dot(data[:,j].reshape(height,1),(Xn[:,j].reshape(len(Xn),1).transpose())))

		# Take new value of W
		W = np.dot(temp1,inv(Xn_Xn_T))

		sum11 = 0
		for j in range(0,width):
			temp2 = sigma_square*(In_M) + np.dot(Xn[:,j].reshape(len(Xn),1),(Xn[:,j].reshape(len(Xn),1).transpose()))
			sum11 = sum11 + mt.pow(LA.norm(data[:,j]),2) - 2*np.dot(np.dot(Xn[:,j].reshape(len(Xn),1).transpose(),W_T),data[:,j].reshape(height,1)) + np.trace(np.dot(np.dot(temp2,W_T),W))

		sigma_square = sum11/(height*width)

		mse = ((old_W-W)**2).mean(axis=None)

		if(mse<error):
			break;
		else:
			print 'iter :- ',i+1,' Mean Squared Error :- ',mse

	print('EM is finished')
	M = np.dot(W_T,W) + sigma_square*np.identity(no_pc)
	In_M = inv(M)
	Xn = np.dot(np.dot(In_M,W_T),data)

	return W,sigma_square,Xn,t_mean,M
